feat: Implement taxonomy sample data generation

yusuf-musleh opened the pull request:
### Description
Implements generating sample taxonomy data to be used for testing and development. Repeating what is already mentioned in the README:

Running this script will do the following:

1. 2 Test Organizations will be created
1. For each organization created, a sample course is created that is populated with a course export that contains a variety of section/subsections/units and components
1. For each of these created organizations the following taxonomies will be created:
    1. a disabled taxonomy with 10 Tags
    1. an enabled flat taxonomy with 5000 Tags
    1. an enabled hierarchical taxonomy with three levels and 4^x tags per level (4 root tags, each with 16 child tags, each with 64 grandchild tags)
    1. a small enabled taxonomy with 2 levels with 2 Tags each
1. A multi org Taxonomy is created and enabled/used by both orgs
1. (Optional) A 4 level Taxonomy containing data obtained from [Open Canada Taxonomy](https://open.canada.ca/data/en/dataset/6093c709-2a0d-4c23-867e-27987a79212c/resource/0a120b15-9708-4d8a-8af2-2431c4540c0b)
1. (Optional) A 3 level Taxonomy containing data obtained from [LightCast Skills Taxonomy](https://docs.google.com/spreadsheets/d/1DA3JfpBE5Krc0daImuu5Y0nsH93PEfdrWRrEa-sR-6k/edit#gid=1319222368)
1. Once the Taxonomies and their Tags have been created, the script will Tag each of the courses along with all the components they contain with 1 of each Tag from the the above

**Note:** This script is designed to be idempotent. Meaning that the end state is the same every time you run it. So if you make modifications to the sample courses on Studio or the Taxonomy data in the shell and run this script again, it will reset all your changes.

### Related Tickets:
- Closes https://github.com/openedx/modular-learning/issues/107

### Testing Instructions
Similar to what is mentioned in the README:
1. To begin, clone this repo inside a directory so it can be accessed from within the devstack (eg: `/edx/src/`)
1. Set the `TAXONOMY_SAMPLE_PATH` variable inside the `generate.py` file to point to the cloned repo inside your devstack/tutor environment, eg: `/edx/src/taxonomy-sample-data`
1. (Optional) If you would not like to include taxonomy data from real world examples, such as:
    - [Open Canada Taxonomy](https://open.canada.ca/data/en/dataset/6093c709-2a0d-4c23-867e-27987a79212c/resource/0a120b15-9708-4d8a-8af2-2431c4540c0b)
    - [LightCast Skills Taxonomy](https://docs.google.com/spreadsheets/d/1DA3JfpBE5Krc0daImuu5Y0nsH93PEfdrWRrEa-sR-6k/edit#gid=1319222368)

    Then set the following flags to `False` in `generate.py` accordingly:

    ```py
    IMPORT_OPEN_CANADA_TAXONOMY = True
    IMPORT_LIGHTCAST_SKILLS_TAXONOMY = True
    ```

1. To run the script, enter the LMS shell (`make lms-shell`) and run the following command:
    ```sh
    python manage.py cms shell < /path/to/taxonomy-sample-data/generate.py
    ```

4. The script should run to completion, you'll notice it tagging the courses + components in the logs of the script

---
Private-ref: [FAL-3516](https://tasks.opencraft.com/browse/FAL-3516)

yusuf-musleh committed:
feat: Add creating sample orgs, courses + populate

yusuf-musleh committed:
feat: Add sample taxonomies generation for orgs

yusuf-musleh committed:
feat: Add sample tag generation for taxonomies

The following sample tags will be generated:
  - 10 Tags for the disabled taxonomy
  - 5000 Tags for the flat taxonomy
  - 4^x Tags with 3 levels for the hierarchical taxonomy
  - 2 Tags with 2 levels for two level taxonomy
  - 5 Tags for the multi org taxonomy

Whenever the generate script is run, it clears the existing tags and
creates fresh new ones

yusuf-musleh committed:
feat: Add tags togenerated courses + components

yusuf-musleh committed:
feat: Add Open Canada Taxonomy data as sample

Include toggle-able Open Canada Taxonomy data in the generate script
from:

https://open.canada.ca/data/en/dataset/6093c709-2a0d-4c23-867e-27987a79212c/resource/0a120b15-9708-4d8a-8af2-2431c4540c0b

yusuf-musleh committed:
feat: Add LightCast Taxonomy data + refactor

yusuf-musleh committed:
docs: Add README.md

bradenmacdonald commented on generate.py:
```diff
@@ -0,0 +1,593 @@
+    create_taxonomy, get_taxonomies_for_org,
+    set_taxonomy_orgs, tag_content_object, get_content_tags,
+    resync_object_tags
+)
```
Hmm, I guess I forgot to point out that we already have an import-export API for taxonomies that can import JSON files? 😬  https://github.com/openedx/openedx-learning/tree/main/openedx_tagging/core/tagging/import_export

Not a big deal, but it would have been nice to re-use the existing import code and JSON file format. No need to change now though, we need to wrap this up.

yusuf-musleh replied:
Oh my bad 😅  didn't know we had that. I can address that in a future ticket for sure.

bradenmacdonald commented on README.md:
````diff
@@ -0,0 +1,41 @@
+
+
+```sh
+source /edx/app/edxapp/edxapp_env && python /edx/app/edxapp/edx-platform/manage.py cms shell < /edx/src/taxonomy-sample-data/generate.py
````
Is `source /edx/app/edxapp/edxapp_env` still necessary? I haven't used devstack in a long time, but I thought that was done for you automatically these days. Certainly on tutor dev environments you can just run `python` or `manage.py...` and it's already set up with the right virtual environment.

yusuf-musleh replied:
You're right, it's not necessary. I added it because it was mentioned in the ticket as an example to run the command. I updated the instructions in the PR + README.

bradenmacdonald commented on generate.py:
```diff
@@ -0,0 +1,593 @@
+MULTI_ORG_TAXONOMY_NAME = "MultiOrgTaxonomy"
+
+IMPORT_OPEN_CANADA_TAXONOMY = False
+IMPORT_LIGHTCAST_SKILLS_TAXONOMY = False
```
Can we just make these True by default? I think we'll generally want to have them.

yusuf-musleh replied:
Sure thing

bradenmacdonald commented on generate.py:
```diff
@@ -0,0 +1,593 @@
+
+user = User.objects.get(email=USER_EMAIL)
+
+TARFILE_PATH = '/edx/src/taxonomy-sample-data/course.g4vmy6n2.tar.gz'
```
`/edx` doesn't exist on Tutor-based installations like my devstack or the sandbox where we'll need to install these. And non-root users can't create `/edx`. So could you tweak this so the `/edx/src/taxonomy-sample-data/` path can be changed via environment variable?

e.g. `TAXONOMY_SAMPLE_PATH=/openedx/requirements/taxonomy-sample-data python manage.py cms shell < $TAXONOMY_SAMPLE_PATH/generate.py`

would be how I would like to run this command on a tutor-based devstack.

yusuf-musleh replied:
So I tried to get it working using env variables, however because of the way the script is run the env variables are not transferred over to the script. I looked into it a bit, but I didn't want to drag it out. So I took an alternate approach of just having to set the variable in the `generate.py` file. So now all you need to run the script after setting the variable is the following:

`python manage.py cms shell < /openedx/requirements/taxonomy-sample-data/generate.py`

bradenmacdonald replied:
I see. OK, no problem. Let's keep it simple!

bradenmacdonald reviewed (commented):
Generally looks great, thanks! I'll try to test this tomorrow; didn't get time today. If you can make the path customizeable in the meantime, that will make it easier for me to test since I don't use devstack.

yusuf-musleh committed:
feat: Remove hardcoded path + Update run command

bradenmacdonald commented on README.md:
````diff
@@ -0,0 +1,40 @@
+1. To run the script, enter the LMS shell (`make lms-shell`) and run the following command:
+    ```sh
+    python manage.py cms shell < /path/to/taxonomy-sample-data/generate.py
+    ```
````
We need to add License information to the README:

This includes a copy of the "[Skills and Competencies Taxonomy Data](https://open.canada.ca/data/en/dataset/6093c709-2a0d-4c23-867e-27987a79212c)" from Employment and Social Development Canada, which is under the [Open Government Licence - Canada](http://open.canada.ca/en/open-government-licence-canada).

This includes a copy of the [Lightcast Open Skills Taxonomy](https://lightcast.io/open-skills), which is "free for individual and not-for-profit use" and is available under the [Lightcast Open Skills Terms of Use](https://lightcast.io/downloads/lightcast-open-skills-license-agreement.pdf).

Also, since you have copied AGPL code from edx-platform, the license of this repo should be AGPL, not GPL.

yusuf-musleh replied:
Updated the repo license, and added these licenses information to the README.

bradenmacdonald commented on generate.py:
```diff
@@ -0,0 +1,599 @@
+multi_org_taxonomy_tags = create_tags_for_multi_org_taxonomy(multi_org_taxonomy)
+
+
+for org in sample_orgs:
```
Could we make the LightCast taxonomy and Canada Skill taxonomies into "all orgs" taxonomies instead? (Just one copy of each instead of two.) Any taxonomy that's coming from an external source doesn't need a separate copy for each org.

yusuf-musleh replied:
Got it, done.

bradenmacdonald reviewed (changes requested):
Looks great and seems to have worked well! A couple notes:

bradenmacdonald commented:
BTW I also had to change the default user name, because there are no standard users on Tutor devstacks. So the first time I ran it, I saw an error about UserDoesNotExist, but I fixed it easily by changing the variable.

yusuf-musleh committed:
docs: Add License information + note on USER_EMAIL

yusuf-musleh committed:
feat: Make LightCast/Canada Skill Taxonomy all org

yusuf-musleh commented:
> BTW I also had to change the default user name, because there are no standard users on Tutor devstacks. So the first time I ran it, I saw an error about UserDoesNotExist, but I fixed it easily by changing the variable.

Oh I see, I thought `edx@example.com` was a universally available user. I added a note in the README for folks to modify if needed.

bradenmacdonald reviewed (approved):
:+1:

- [x] I tested this: imported it into my devstack
- [x] I read through the code
- [x] I checked for accessibility issues: n/a
- [x] Includes documentation: README

bradenmacdonald commented:
@pomegranited @rpenido @ChrisChV Please check out this nice work from @yusuf-musleh! I'd like everyone working on the taxonomy project to import these taxonomies/courses/tags into their devstack please. In FAL-3531 we will soon put these onto the sandbox too.
